FibonacciHeap.java

```diff
@@ -88,9 +88,51 @@ public void deleteMin() {
 
 	}
 
+	/**
+	 * 
+	 * Consolidates the root list so at most one tree has each rank.
+	 * 
+	 */
 	private void consolidate() {
-		// TODO: By Yonatan.
-		return;
+		HeapNode[] buckets = new HeapNode[t_num];
+		HeapNode current = first;
+		do {
+			int rank = current.rank;
+			HeapNode merged = current;
+			// Recursively merge trees until there are no more bucket conflicts
+			while (buckets[rank] != null) {
+				// Merge with bucket
+				merged = merge(merged, buckets[rank]);
+				// Empty previous bucket
+				buckets[rank++] = null;
+			}
+			buckets[rank] = merged;
+
+			current = current.next;
+		} while (current != first);
+	}
+
+	/**
+	 * 
+	 * Merges two trees of the same rank. Returns the merged tree.
+	 * 
+	 */
+	private HeapNode merge(HeapNode x, HeapNode y) {
+		HeapNode parent = x.key <= y.key ? x : y;
+		HeapNode child = x.key <= y.key ? y : x;
+
+		// Remove child from the root list
+		removeFromLinkedList(child);
+		// Attatch child to parent
+		if (parent.child == null) {
+			parent.child = child;
+			child.next = child;
+			child.prev = child;
+		} else {
+			addToLinkedList(parent.child, child);
+		}
+
```

Comment:
Increase parent rank as it just received a new child
```suggestion
parent.rank++;
```